File: src/chain/examples/idvalidation/IdValidator.js
'use strict';

/** RNF3.1: Chain of Responsibility. Perfil habilitado solo si todo pasa. */
class IdValidator {
  constructor() { this.next = null; }

  linkWith(next) {
    this.next = next;
    return next;
  }

  validate(doc) {
    if (!this.check(doc)) {
      console.log(`  X rechazado en ${this.name()}`);
      return false;
    }
    console.log(`  OK ${this.name()}`);
    return this.next ? this.next.validate(doc) : true;
  }

  check(_doc) { throw new Error('not implemented'); }
  name()      { throw new Error('not implemented'); }
}

module.exports = IdValidator;

File: src/chain/examples/idvalidation/IdValidator.java
package chain.examples.idvalidation;

/**
 * RNF3.1: caretaker profile stays hidden until every check passes.
 * Chain of Responsibility.
 */
public abstract class IdValidator {

    protected IdValidator next;

    public IdValidator linkWith(IdValidator next) {
        this.next = next;
        return next;
    }

    public final boolean validate(IdDocument doc) {
        if (!check(doc)) {
            System.out.println("  X rechazado en " + name());
            return false;
        }
        System.out.println("  OK " + name());
        return next == null || next.validate(doc);
    }

    protected abstract boolean check(IdDocument doc);
    protected abstract String name();
}
